### PR TITLE
Complete callables without parens if present in locals_without_parens

### LINKED ISSUE #601: Functions listed in `locals_without_parens` should not have parens in their completions
This has been a minor annoyance for a while, but when we complete `assert`, the completion comes in as `assert(assertion)`, but it should be `asert assertion`.

### PR DESCRIPTION
**Pull request by zachallaun**
This is scratching an itch I had while I was doing some work on another library.

Fixes #601

### CODE REVIEW

**Comment by scohen** [comment]
This fixes #601 right? Awesome!

File: apps/remote_control/lib/lexical/remote_control/completion.ex [hunk 1]
```diff
@@ -18,14 +19,46 @@ defmodule Lexical.RemoteControl.Completion do
...
+       when struct in [Candidate.Function, Candidate.Macro] do
+    atom_name = String.to_atom(candidate.name)
+
+    suppress_parens? =
+      locals_without_parens
+      |> Keyword.take([atom_name])
+      |> Keyword.values()
+      |> Enum.any?(&suppress_parens?(&1, candidate.arity))
+
+    if suppress_parens? do
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/completion.ex:49]: this is equivalent to `%{candidate | parens?: not suppress_parens}`

File: apps/remote_control/lib/lexical/remote_control/completion/candidate.ex [hunk 2]
```diff
@@ -4,7 +4,18 @@ defmodule Lexical.RemoteControl.Completion.Candidate do
...
+      :argument_names,
+      :arity,
+      :name,
+      :origin,
+      :type,
+      :visibility,
+      :spec,
+      :summary,
+      :metadata,
+      parens: true
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/completion/candidate.ex:17]:
```suggestion
      parens?: true
```

File: apps/remote_control/lib/lexical/remote_control/completion/candidate.ex [hunk 3]
```diff
@@ -21,7 +32,17 @@ defmodule Lexical.RemoteControl.Completion.Candidate do
...
+    defstruct [
+      :argument_names,
+      :arity,
+      :metadata,
+      :name,
+      :origin,
+      :spec,
+      :summary,
+      :type,
+      parens: true
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/completion/candidate.ex:44]:
```suggestion
      parens?: true
```

File: apps/remote_control/lib/lexical/remote_control/completion/candidate.ex [hunk 4]
```diff
@@ -38,7 +59,17 @@ defmodule Lexical.RemoteControl.Completion.Candidate do
...
+    defstruct [
+      :argument_names,
+      :arity,
+      :name,
+      :origin,
+      :type,
+      :visibility,
+      :spec,
+      :metadata,
+      parens: true
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/completion/candidate.ex:71]:
```suggestion
      parens?: true
```

File: apps/remote_control/lib/lexical/remote_control/completion/candidate.ex [hunk 5]
```diff
@@ -55,7 +86,18 @@ defmodule Lexical.RemoteControl.Completion.Candidate do
...
+      :argument_names,
+      :arity,
+      :doc,
+      :metadata,
+      :type,
+      :name,
+      :origin,
+      :signature,
+      :spec,
+      parens: true
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/completion/candidate.ex:99]:
```suggestion
      parens?: true
```

File: apps/remote_control/lib/lexical/remote_control/completion.ex [hunk 6]
```diff
@@ -18,14 +19,46 @@ defmodule Lexical.RemoteControl.Completion do
...
   end
 
+  defp from_elixir_sense(suggestion, locals_without_parens) do
+    suggestion
+    |> Candidate.from_elixir_sense()
+    |> maybe_suppress_parens(locals_without_parens)
+  end
+
+  defp maybe_suppress_parens(%struct{} = candidate, locals_without_parens)
+       when struct in [Candidate.Function, Candidate.Macro] do
```
Comment (scohen) [inline apps/remote_control/lib/lexical/remote_control/completion.ex:40]:
```suggestion
       when struct in [Candidate.Function, Candidate.Macro, Candidate.Typespec] do
```

**Review by scohen — COMMENTED** [review]
This is awesome. I feel this pain every time I type `assert` and it adds `assert(`
Can we have some tests?

**Comment by zachallaun** [comment]
@scohen Thanks for the feedback! Agreed on tests; it does need more.

I'm actually not 100% sure that `assert` is going to complete without parens because the ExUnit assertions aren't actually in `locals_without_parens`. That makes me wonder what the formatter rules are for macros in built-in packages. Do they all exclude parens?

**Comment by scohen** [comment]
I tried this branch, and it still adds parens to all the exunit macros. I think this is because it only checks the `locals_withut_parens` and not `import_deps`.

**Comment by zachallaun** [comment]
> I tried this branch, and it still adds parens to all the exunit macros. I think this is because it only checks the `locals_withut_parens` and not `import_deps`.

See my message above -- the ExUnit macros aren't included for whatever reason, I think we need to special-case some things. I'm going to look into how the Elixir formatter decided whether to put parens around calls.

Re: `import_deps`, they're expanded automatically into `locals_without_parens`. Here's what it looks like for a library of mine, for instance:

```elixir
iex(3)> Mix.Tasks.Format.formatter_for_file("lib/mneme.ex")
{#Function<14.79746710/1 in Mix.Tasks.Format.find_formatter_for_file/2>,
 [
   sigils: [],
   inputs: ["{mix,.formatter}.exs", "{config,lib}/**/*.{ex,exs}", "test/**/*.{ex,exs}",
    "test_integration/**/*.{ex,exs}", "examples/*.{ex,exs}"],
   locals_without_parens: [
     all: :*,
     check: 1,
     check: 2,
     property: 1,
     property: 2,
     auto_assert: :*,
     auto_assert_raise: :*,
     auto_assert_receive: :*,
     auto_assert_received: :*
   ],
   export: [
     locals_without_parens: [
       auto_assert: :*,
       auto_assert_raise: :*,
       auto_assert_receive: :*,
       auto_assert_received: :*
     ]
   ],
   import_deps: [:stream_data],
   plugins: [Styler],
   line_length: 98
 ]}
```

`all`, `check` and `property` come from `:stream_data`.

**Comment by zachallaun** [comment]
Looks like the formatter [special-cases](https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/code/formatter.ex#L60) a bunch of things. We should likely copy this list and can set `parens?: false` for those.

**Comment by zachallaun** [comment]
@scohen Still need to add more tests but give things a try now.

**Comment by scohen** [comment]
@zachallaun much better. This is good to go when the tests are done.

**Comment by zachallaun** [comment]
@scohen Pushed some additional tests. Are there other cases you'd like to see?

**Review by scohen — APPROVED** [review]
consider my comment, and then you're good to go.

File: apps/remote_control/lib/lexical/remote_control/completion.ex [hunk 7]
```diff
@@ -4,10 +4,96 @@ defmodule Lexical.RemoteControl.Completion do
   alias Lexical.Document
   alias Lexical.Document.Position
   alias Lexical.RemoteControl
+  alias Lexical.RemoteControl.CodeMod.Format
   alias Lexical.RemoteControl.Completion.Candidate
 
   import Document.Line
 
+  @built_in_locals_without_parens [
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/remote_control/lib/lexical/remote_control/completion.ex]: Makes sense! Just pushed that change -- if you could confirm that it does what you were thinking, I think we're good to go.

**Pull request merged**